flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,1683 @@
+                # Half spaces
+                self.vertical(0).left_half_space(),
+                self.half_circle(0, 2).left_half_space(),
+                ]
+
+    def _element_constructor_(self, x):
+        r"""
+        Return ``x`` as an element of the hyperbolic plane.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: H = HyperbolicPlane(QQ)
+            sage: H(H.an_element()) in H
+            True
+
+        Base ring elements can be converted to ideal points::
+
+            sage: H(1)
+            1
+
+        The point at infinity in the half plane model can be written directly::
+
+            sage: H(oo)
+            ∞
+
+        Complex numbers in the upper half plane can be converted to points in
+        the hyperbolic plane::
+
+            sage: H(I)
+            I
+
+        Elements can be converted between hyperbolic planes with compatible base rings::
+
+            sage: HyperbolicPlane(AA)(H(1))
+            1
+
+        """
+        if x.parent() is self:
+            return x
+
+        from sage.all import Infinity
+        if x is Infinity:
+            return self.infinity()
+
+        if x in self.base_ring():
+            return self.real(x)
+
+        if isinstance(x, HyperbolicConvexSet):
+            return x.change_ring(self.base_ring())
+
+        from sage.categories.all import NumberFields
+        if x.parent() in NumberFields():
+            K = x.parent()
+
+            from sage.all import I
+            if I not in K:
+                raise NotImplementedError("cannot create a hyperbolic point from an element in a number field that does not contain the imaginary unit")
+
+            return self.point(x.real(), x.imag(), model="half_plane")
+
+        raise NotImplementedError("cannot create a subset of the hyperbolic plane from this element yet.")
+
+    def base_ring(self):
+        r"""
+        Return the base ring over which objects in the plane are defined.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().base_ring()
+            Rational Field
+
+        """
+        return self._base_ring
+
+    def infinity(self):
+        r"""
+        Return the point at infinity in the Poincaré half plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().infinity()
+            ∞
+
+        """
+        return self.projective(1, 0)
+
+    def real(self, r):
+        r"""
+        Return the ideal point ``r`` on the real axis in the Poincaré half
+        plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().real(-2)
+            -2
+
+        """
+        return self.projective(r, 1)
+
+    def projective(self, p, q):
```

Comment:
I think that projective can refer to at least to three things
- `[p: q]` in P^1(R) (boundary of the Poincare upper half plane)
- `[p: q]` in P^1(C) (Poincare upper half plane)
- `[p: q: r]` in P^2(R) (Klein model)

Reply:
Ok. Then it should get a `model` parameter.